work_init: React-query

erastian opened the pull request:
- reworked tree and router
- refactored imports
- fixed Card and newsFeed layout
- invented Icons component and removed dependent svg files
- created ArticlePage component
- added in router suspended element
- fixed Button component IProps
- refactored Button component
- refactored IconSet component
- added Axios lib
- built boilerplate modules for query with axios

erastian committed:
work_init: React-query

- reworked tree and router
- refactored imports

erastian committed:
work_init: React-query

- fixed Card layout
- fixed newsFeed layout

mdziarkach commented on vite.config.ts:
```diff
@@ -1,7 +1,18 @@
   plugins: [react()],
+  resolve: {
+    alias: {
+      '~app': path.resolve('src/app'),
```
'@': path.resolve('src'),

erastian replied:
Done but by default @ goes to root folder. MB somehow it needs to be synced with tsconfig.json?

mdziarkach replied:
sure, you need to sync it

Mikhail1992 replied:
looks like you didn't understand me :)
if you use this line 
```js
'@': path.resolve('src/'),
// or
'~': path.resolve('src/'),
```
you don't need rest part of aliases

erastian replied:
Right now using ~ is comfortable

mdziarkach commented on index.html:
```diff
@@ -8,6 +8,6 @@
   <body>
     <div id="root"></div>
-    <script type="module" src="/src/main.tsx"></script>
+    <script type="module" src="/src/app/index.tsx"></script>
```
you don't need to change it

erastian replied:
Yeap. Moved it only for to be clear and less files

mdziarkach commented on package.json:
```diff
@@ -23,6 +26,7 @@
     "eslint": "^8.46.0",
     "eslint-plugin-react-hooks": "^4.6.0",
     "eslint-plugin-react-refresh": "^0.4.3",
+    "path": "^0.12.7",
```
I did't fint where you use it

erastian replied:
In vite.config.ts

mdziarkach commented on src/app/providers/index.tsx:
```diff
@@ -0,0 +1,16 @@
+        </BrowserRouter>
+      </QueryClientProvider>
+  )
+}
```
looks like src/app/App.tsx file

erastian replied:
Yes, it is.

mdziarkach commented on src/entities/card/Card.tsx:
```diff
@@ -0,0 +1,37 @@
+
+export function Card({ articleData, pinned = false }: CardInterface) {
+  return (
+      <div className={ `${ styles.card } ${ pinned ? `${ styles.pinned } d-col-3 m-col-6` : 'd-col-12 m-col-12' }` }>
```
very complicated
maybe better to have 2 components?

erastian replied:
DRY?)

mdziarkach replied:
no, it doesn't apply here

erastian replied:
Done

mdziarkach commented on src/entities/card/Card.tsx:
```diff
@@ -0,0 +1,37 @@
+            <div className="counters">
+                <span className="comments"><img src={ iconMessages }
+                                                alt="icon"></img> { articleData._count.comments }</span>
+              <span className="views"><img src={ iconUsers } alt="icon"></img> 500</span>
```
don't use img for icons please
use `<svg />`

erastian replied:
Convert all iconset to component or what do you mean?

mdziarkach replied:
create please svg sprite and then you can use this sprite in `SVGIcon` component for example
https://dev.to/mbarzeev/using-svg-sprites-in-a-react-app-477d

erastian replied:
Done. All icons was converted to sprite file

mdziarkach commented on src/entities/card/card.module.css:
```diff
@@ -31,10 +31,13 @@
     line-height: normal;
 }
+
+.card :global .details .category a {
```
very complicated selector. I'm sure it possible to simplify it

erastian replied:
Done. Used react styles declaration

mdziarkach commented on src/entities/topMenu/TopMenu.tsx:
```diff
@@ -1,10 +1,11 @@
+import { PAGE_PATH } from "~shared/lib/router/paths.ts";
 
-function TopMenu() {
+export function topMenu() {
```
TopMenu was right

erastian replied:
Done

mdziarkach commented on src/shared/api/Api.ts:
```diff
@@ -0,0 +1,24 @@
+export interface ArticleDTO {
```
IArticle

erastian replied:
Done

mdziarkach commented on src/shared/lib/router/paths.ts:
```diff
@@ -0,0 +1,13 @@
+  login: '/login',
+  register: '/register',
+  profile: '/profile',
+  article: {
```
article: "/article/:url"

erastian replied:
The current implementation is quite comfortable to use

mdziarkach commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,27 @@
+import styles from './Button.module.css';
+import { ReactNode } from "react";
+
+type ButtonProps = {
```
describe interfaces correctly please https://stackoverflow.com/questions/40731352/extending-html-elements-in-react-and-typescript-while-preserving-props

```ts
interface IProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {}
```
this interface already include what you need

erastian replied:
Done

mdziarkach commented on src/shared/ui/fullPageWrapper/full-page-wrapper.module.css:
```diff
@@ -0,0 +1,3 @@
+.wrapper {
+    min-height: 22rem;
```
please don't use rem for page size
rem depends on base font size and it will work unpredictable on mobile devices

erastian replied:
Its just a spacer for situations where I need to use Spinner component. Just a decoration. Nothing more.

mdziarkach replied:
then you can use viewport

erastian replied:
Nope, if I use spinner in topNews and newsFeed separately)

mdziarkach commented on src/widgets/hotNews/HotNews.tsx:
```diff
@@ -0,0 +1,39 @@
+import { Spinner } from "~shared/ui/spinner";
+
+export function HotNews() {
+  const demoData = {
```
at least move it out of component please

mdziarkach replied:
not fixed

erastian replied:
fixed

mdziarkach commented on src/widgets/newsFeed/NewsFeed.tsx:
```diff
@@ -0,0 +1,37 @@
+export function NewsFeed() {
+  const { isLoading, data: newsFeed, error } = useQuery({
+    queryKey: [ 'newsFeed' ],
+    queryFn: async () => {
```
looks like it should be in entities/article/model

erastian replied:
It will be. It just a draft

mdziarkach commented on src/widgets/newsFeed/api/feedApi.tsx:
```diff
@@ -0,0 +1,40 @@
+    const responce = await fetch('http://localhost:8000/articles')
+    return await responce.json()
+  }
+}
```
need to refactor it
1. All requests should be somewhere in api
2. Work with data in entities
3. Infinity scroll should include limit (default number of items)

erastian replied:
Also that draft too. It will be placed in shared

mdziarkach commented on tsconfig.json:
```diff
@@ -18,7 +18,30 @@
+      "~entities/*": ["./entities/*"],
+
+      "~shared": ["./shared"],
+      "~shared/*": ["./shared/*"]
```
"~app/*": ["app/*"],
"~processes/*": ["processes/*"],
"~pages/*": ["pages/*"],
...

try it please

erastian replied:
Done

erastian committed:
work_init: React-query

- invented Icons component and removed dependent svg files
- created ArticlePage component
- added in router suspended element
- fixed Button component IProps

mdziarkach commented on src/shared/ui/button/Button.tsx:
```diff
@@ -1,7 +1,7 @@
 
-type ButtonProps = {
+type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
   style?: string;
```
style, type, children, onClick - already exist into Base type

erastian replied:
Already discuss about this

mdziarkach commented on src/entities/card/card.module.css:
```diff
@@ -19,14 +19,14 @@
 }
 
-.card :global .details {
+.card .details {
```
now, when you already use css modules you don't need nesting
so you can replace 
```css
.card .details {}
```
to
```css
.details {}
```

erastian replied:
Yeap. But its not a problem with shallow nesting. Im using it just for visual comfort. But anyway, thanks for the reminder)

erastian committed:
work_init: React-query

- refactored Button component
- refactored IconSet component
- added Axios lib
- built boilerplate modules for query with axios

mdziarkach commented on package.json:
```diff
@@ -10,9 +10,14 @@
+    "react-query": "^3.39.3",
+    "react-redux": "^8.1.2",
+    "react-router-dom": "^6.15.0",
+    "vite-plugin-svgr": "^4.1.0"
```
looks like devDependency

erastian replied:
fixed

mdziarkach commented:
are you sure you need this image src/assets/img-not-found.svg? you can add this placeholder and css styles without img

mdziarkach commented on src/entities/article/api/articleApi.ts:
```diff
@@ -0,0 +1,103 @@
+    getNextPageParam: (lastPage, pages) => {
+      if (lastPage.length < limit) return null;
+
+      const nextPageParam = lastPage.length ? pages.length * limit : null;
```
if lastPage.length == 0 then this condition `if (lastPage.length < limit) return null;` will catch it

```js
return pages.length * limit
```

erastian replied:
The main idea here about "if last page length less than limit we do not have next page". Unfortunately, in cases when we are have number of topics equal total topics divided on limit without reminder, next page also will be available, but will contains nothing. IDK how to avoid this behavior without extra work with BE part

mdziarkach commented on src/entities/article/api/articleApi.ts:
```diff
@@ -0,0 +1,103 @@
+import { Api, RequestParams } from "~shared/api/Api.ts";
+import { useInfiniteQuery } from "react-query";
+
+export interface IArticle {
```
move please types to another file

erastian replied:
moved

mdziarkach commented on src/entities/card/Card.tsx:
```diff
@@ -0,0 +1,43 @@
+import { articleApi } from "~entities/article";
+import Icon from "~shared/ui/icon";
+
+type CardInterface = {
```
it will be simpler to call this type - Props in each component with same name

erastian replied:
fixed

mdziarkach commented on src/entities/card/Card.tsx:
```diff
@@ -0,0 +1,43 @@
+        </div>
+      </div>
+  );
+}
```
split it into 2 components

erastian replied:
Done

mdziarkach commented on src/entities/card/card.module.css:
```diff
@@ -19,54 +19,57 @@
 }
 
-.card :global .details {
+.card .details {
```
you use css modules and can use simple selectors

```css
.details {
```

erastian replied:
done

mdziarkach commented on src/entities/card/card.module.css:
```diff
@@ -19,54 +19,57 @@
+.card .details .category a {
+    color: var(--color-primary-accent);
+}
+.card .details .counters {
```
remove nesting, please
```css
.counters {
```

erastian replied:
done

mdziarkach commented on src/entities/topMenu/TopMenu.tsx:
```diff
@@ -1,10 +1,11 @@
+          <NavLink to={ PAGE_PATH.root }>All</NavLink>
         </li>
         <li>
           <NavLink to='/places'>Places</NavLink>
```
```jsx
<NavLink to={makeURL(PAGE_PATH..., { category: "places"})}>Places</NavLink>
```

erastian replied:
Fixed

mdziarkach commented on src/entities/userMenu/UserMenu.tsx:
```diff
@@ -0,0 +1,16 @@
+
+export function UserMenu() {
+  return (
+      <div className='user-menu'>
```
use, please one styling approach
I see that in a lot of places you already use css modules

erastian replied:
Fixed

mdziarkach commented on src/pages/article-page/article.tsx:
```diff
@@ -0,0 +1,19 @@
+      <FullPageWrapper>
+        <div className="">
+          <h3>{ articleName }</h3>
+          <p><span>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Ab cupiditate dicta fuga, illo in obcaecati sequi similique veritatis. Amet autem culpa ea facere impedit libero natus non quam, reprehenderit sunt?</span><span>Aliquam atque culpa eligendi exercitationem expedita natus nisi nobis! Cumque dolore est fugit officia officiis, repellendus similique temporibus. Aliquid beatae ...
```
you please mock data if you need temporary properties in your components

erastian replied:
Refactored with real data from API

mdziarkach commented on src/pages/article-page/article.tsx:
```diff
@@ -0,0 +1,19 @@
+import { FullPageWrapper } from "~shared/ui/fullPageWrapper";
+import useDocumentTitle from "~shared/document-title";
+//
+// type IArticleProps = {
```
remove commented code please

erastian replied:
removed

mdziarkach commented on src/pages/home/HomePage.tsx:
```diff
@@ -0,0 +1,15 @@
+  return (
+      <FullPageWrapper>
+        <HotNews/>
+        <NewsFeed query={{ limit: 10, offset: 0}}/>
```
```jsx
<NewsFeed />
```
move limit in configuration please

erastian replied:
moved

mdziarkach commented on src/pages/page-404/Page404.tsx:
```diff
@@ -1,15 +1,14 @@
-      <div className='container'>
+      <FullPageWrapper>
         <h3 className="text-center mb-3">Error 404: Page Not Found</h3>
         <p className="text-center">Please, try <NavLink style={{textDecoration: 'underline'}} to='/'>starting from the beginning</NavLink>.</p>
```
```jsx
style={{textDecoration: 'underline'}}
```
let's use one common styling approach

erastian replied:
fixed

mdziarkach commented on src/shared/api/Api.ts:
```diff
@@ -0,0 +1,52 @@
+
+
+export class Api {
+  static articles = {
```
for what do you need this class? looks like class can be removed

erastian replied:
Done in next PR

mdziarkach commented on src/shared/api/Api.ts:
```diff
@@ -0,0 +1,52 @@
+
+export type RequestParams = Omit<FullRequestParams, 'body' | 'method' | 'query' | 'path'>;
+
+export interface FullRequestParams extends Omit<RequestInit, 'body'> {
```
for what do you need this type?

erastian replied:
Removed in next PR

mdziarkach commented on src/shared/api/base.ts:
```diff
@@ -0,0 +1,75 @@
+import { API_URL } from "~shared/config";
+import axios, { AxiosInstance, AxiosResponse, AxiosRequestConfig } from "axios";
+
+class ApiInstance {
```
for what is it need?

```ts
export const request = axios.create({
      baseURL: API_URL,
      headers: {
        "Content-Type": "application/json",
      },
    });
```

this should be enough

erastian replied:
Done in next PR

mdziarkach commented on src/shared/api/index.ts:
```diff
@@ -0,0 +1,13 @@
+  }
+})
+
+export { newsApi };
```
seems to me this file can be removed

erastian replied:
removed

mdziarkach commented on src/shared/config/route-paths.ts:
```diff
@@ -0,0 +1,13 @@
+  login: '/login',
+  register: '/register',
+  profile: '/profile',
+  article: {
```
we had an agreement that you should create a function which will set params in urls and here we can leave only simple URLs]
```
article: '/article/:url'
```

erastian replied:
The current implementation is quite comfortable to use

mdziarkach commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,27 @@
+type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
+  size?: '' | 'semi' | 'large';
+  variant?: string;
+  onClick?: () => void;
```
remove onClick please
this is already exists in ButtonHTMLAttributes

erastian replied:
Without declaration TS very angry on this "onClick" function

mdziarkach commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,27 @@
+
+export function Button(props: ButtonProps) {
+  const {
+    className,
```
better to remove `className`
all styles should be controllable from Button component, not from outside

erastian replied:
I use it for adding global classes like margins, flex, paddings etc
Without declaration I cant put className in component

mdziarkach commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,27 @@
+
+type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
+  size?: '' | 'semi' | 'large';
+  variant?: string;
```
describe all available variants please
it should be enum
and add default value in props please

erastian replied:
Done

mdziarkach commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,27 @@
+import React from "react";
+
+type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
+  size?: '' | 'semi' | 'large';
```
usually 'xs', 'sm', 'md', 'lg'
remove empty string and add default value in props

erastian replied:
removed & added

mdziarkach commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,27 @@
+    onClick,
+    disabled,
+    children
+  } = props;
```
```jsx
export function Button({  variant, size = 'md', children, ...restProps }: ButtonProps) {
....
<button className={cn(styles.button, {
  [styles[size]]: !!size,
  [styles[variant]: !!variant
})} {...restProps}>{ children }</button>
```

mdziarkach commented on src/shared/ui/button/button.module.css:
```diff
@@ -38,6 +38,11 @@
+    border: none;
+}
+
+.button:global(.primary):disabled:hover {
```
simplify styles please

```css
.primary:disabled:hover {
```

erastian replied:
Done

mdziarkach commented on src/shared/ui/fullPageWrapper/FullPageWrapper.tsx:
```diff
@@ -0,0 +1,12 @@
+export function FullPageWrapper(props: FullPageWrapperProps){
+  const { children } = props;
+
+  return <div className={`container ${styles['wrapper']}`}>{children}</div>;
```
```jsx
return <div className={cn(styles.container, styles.wrapper)}`}>{children}</div>;
```

erastian replied:
Fixed

mdziarkach commented on src/shared/ui/spinner/Spinner.tsx:
```diff
@@ -0,0 +1,12 @@
+
+export function Spinner() {
+  return (
+      <div className={styles['circular']}>
```
<div className={styles.circular}>

erastian replied:
fixed

mdziarkach commented on src/widgets/hotNews/HotNews.tsx:
```diff
@@ -0,0 +1,39 @@
+  }
+  return (
+
+      <div className={`${styles.hotNews} flex-grid`}>
```
<div className={cn(styles.hotNews, 'flex-grid')}>

erastian replied:
Fixed

mdziarkach commented on src/widgets/newsFeed/NewsFeed.tsx:
```diff
@@ -0,0 +1,45 @@
+      <div className={ styles.feedWrapper }>
+
+        { status === 'success' ? items.map((article: articleApi.IArticle) => (
+            <Card articleData={ article } key={ article.id }/>)) : 'No data' }
```
which statuses can be in `status`? if only `loading`, `error` and `success` then you already checked `loading` and `error` above and it doesn't make sense to check `success` status here

better to check this logic 
```jsx
if (items.length === 0) {
  return <div className={ styles.feedWrapper }>No data</div>
}

return (
      <div className={ styles.feedWrapper }>
        {items.map((article: articleApi.IArticle) => (Card articleData={ article } key={ article.id }/>)}
        ...
```

erastian replied:
fixed

mdziarkach commented:
looks like you forgot to push your changes

erastian committed:
work_init: React-query

- refactored Button component
- refactored IconSet component
- split Card to 2 components
- fixed global CSS:GRID
- cleaned up API
- added interactivity to Article page
- added ArticleControl component
- a lot of small fixes

erastian committed:
work_init: React-query

- return package-lock.json

mikhail-dziarkach committed:
Update infinity query

mikhail-dziarkach committed:
Merge branch 'work_react-query' of github.com:erastian/REACT_news_client into work_react-query
